datasource/netcdf_groundwater.py

```diff
@@ -75,48 +152,120 @@ def line_type_of(self, line_idx):
+        # gr.nodes.filter(id=100)
+        second = first.filter(id=object_id)
+        # gr.nodes.filter(id=100).timeseries(indexes=slice(None))
+        third = second.timeseries(indexes=slice(None))
+        # get the timeserie of the nc_variable
+        self.vals = getattr(third, nc_variable)
+        # flatten numpyarray
+        self.vals = self.vals.flatten()
+
     def get_timeseries(
             self, object_type, object_id, nc_variable, fill_value=None):
-        return self.temporary_get_timeseries_impl(
-            object_type, object_id, nc_variable, fill_value)
 
-    def temporary_get_timeseries_impl(
-            self, object_type, object_id, nc_variable, fill_value=None):
-        # TODO: this is a crappy but working implementation, replace it by a
-        # better one in the future (e.g. using threedigrid)!
-        import numpy as np
-        if object_type in ['nodes', 'flowlines', 'pumplines']:
-            object_id -= 1
-        if object_type == 'nodes':
-            if object_id < self.nMesh2D_nodes:
-                # it's 2d
-                nc_variable = self.PREFIX_2D + nc_variable
-                netcdf_id = object_id
-            else:
-                # it's 1d
-                nc_variable = self.PREFIX_1D + nc_variable
-                netcdf_id = object_id - self.nMesh2D_nodes
-        elif object_type == 'flowlines':
-            if object_id < self.nMesh2D_lines:
-                # it's 2d
-                nc_variable = self.PREFIX_2D + nc_variable
-                netcdf_id = object_id
-            else:
-                # it's 1d
-                nc_variable = self.PREFIX_1D + nc_variable
-                netcdf_id = object_id - self.nMesh2D_lines
-        elif object_type == 'pumplines':
-            nc_variable = self.PREFIX_1D + nc_variable
-            netcdf_id = object_id
-        else:
-            raise NotImplementedError("TODO")
+        self.gr = self.gridadmin_result
 
-        vals = self.ds.variables[nc_variable][:, netcdf_id]
+        # determine if layer is a not_schematized (e.g nodes, pumps)
+        if get_layer_source(object_type) == 'result':
+            if object_type == 'pumplines':
+                nc_variable = 'Mesh1D_q_pump'
+            self.get_timeseries_result_layer(
+                object_type, object_id, nc_variable)
+        elif get_layer_source(object_type) == 'schematized':
+            if object_type == 'v2_pumpstation':
+                nc_variable = 'Mesh1D_q_pump'
+            self.get_timeseries_schematisation_layer(
+                object_type, object_id, nc_variable)
 
         # Zip timeseries together in (n,2) array
-        if fill_value is not None and type(vals) == np.ma.core.MaskedArray:
-            vals = vals.filled(fill_value)
-        return np.vstack((self.timestamps, vals)).T
+        if fill_value is not None and type(self.vals) == \
+                np.ma.core.MaskedArray:
+            self.vals = self.vals.filled(fill_value)
```

Comment:
Use functions that have inputs and outputs, don't put return values on `self.vals` (see other comments). It is very confusing to follow how `self.vals` gets modified or set just from this context.